src/protocol/stream_decoder.rs

```diff
@@ -0,0 +1,342 @@
+use crate::{
+    models::ERC20Token,
+    protocol::{
+        errors::InvalidSnapshotError,
+        models::{ProtocolComponent, TryFromWithBlock},
+        state::ProtocolSim,
+    },
+};
+use ethers::prelude::H160;
+use futures::{Stream, StreamExt};
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    str::FromStr,
+    sync::Arc,
+};
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use futures::future::BoxFuture;
+use thiserror::Error;
+use tokio::sync::mpsc::Receiver;
+use tokio::sync::RwLock;
+use tokio_stream::wrappers::ReceiverStream;
+use tracing::{debug, warn};
+use tycho_client::feed::{synchronizer::ComponentWithState, FeedMessage, Header};
+use tycho_core::Bytes;
+use crate::evm::tycho_models::Block;
+use crate::protocol::uniswap_v2::state::UniswapV2State;
+use crate::protocol::uniswap_v3::state::UniswapV3State;
+
+#[derive(Error, Debug)]
+pub enum StreamDecodeError {
+    #[error("{0}")]
+    Fatal(String),
+}
+
+#[derive(Debug)]
+pub struct BlockUpdate {
+    pub block_number: u64,
+    /// The current state of all pools
+    pub states: HashMap<String, Box<dyn ProtocolSim>>,
+    /// The new pairs that were added in this block
+    pub new_pairs: HashMap<String, ProtocolComponent>,
+    /// The pairs that were removed in this block
+    pub removed_pairs: HashMap<String, ProtocolComponent>,
```

Comment:
Question: why do we need to track new/removed pools when we have a Hashmap of all the current pools anyway?